cf-agent/simulate_mode.c

```diff
@@ -669,3 +671,166 @@ bool DiffChangedFiles(StringSet **audited_files)
+        return false;
+    }
+
+    StringMap *installed = StringMapNew();
+    StringMap *removed = StringMapNew();
+    char *line;
+    while ((line = GetCsvLineNext(csv_file)) != NULL)
+    {
+        Seq *fields = SeqParseCsvString(line);
+        free(line);
+        if ((fields == NULL) || (SeqLength(fields) != 4))
+        {
+            Log(LOG_LEVEL_ERR, "Invalid package operation record: '%s'", line);
+            SeqDestroy(fields);
+            continue;
+        }
+
+        /* See RecordPkgOperationInChroot() */
+        const char *op       = SeqAt(fields, 0);
+        const char *pkg_name = SeqAt(fields, 1);
+        const char *pkg_ver  = SeqAt(fields, 2);
+        const char *pkg_arch = SeqAt(fields, 3);
+
+        /* These two must always be set properly. */
+        assert(!NULL_OR_EMPTY(op));
+        assert(!NULL_OR_EMPTY(pkg_name));
+
+        /* We need to have a key for the map encoding package name and
+         * architecture. Let's use the sequence "-_-" as a separator to avoid
+         * collisions with possible weird package names. */
+        char *name_arch;
+        xasprintf(&name_arch, "%s-_-%s", pkg_name, pkg_arch ? pkg_arch : "");
+
+        if (*op == CHROOT_PKG_OPERATION_CODE_PRESENT)
+        {
+            /* 'present' operation means that the package was requested to be present and it already
+             * was, so installation didn't happen. However, package operations are not properly
+             * simulated, so the package may have been simulate-removed before and the 'present'
+             * operation would in reality mean the package would be installed back. It's only
+             * recorded as a 'present' operation because the removal doesn't happen in simulate mode
+             * and so the package is still seen as present in the system (package cache).
+             *
+             * This means that a 'present' operation after 'remove' operation results in no
+             * difference (the package would be installed back) so the potential message about the
+             * removal should be removed. */
+            StringMapRemove(removed, name_arch);
+        }
+        else if (*op == CHROOT_PKG_OPERATION_CODE_ABSENT)
+        {
+            /* The same logic applies here for an originally absent package that is installed and
+             * then reported as absent again. No diff to report, just remove the message about the
+             * package installation. */
+            StringMapRemove(installed, name_arch);
+        }
+        else if (*op == CHROOT_PKG_OPERATION_CODE_INSTALL)
+        {
+            /* TODO: check version */
+            bool existed = StringMapRemove(removed, name_arch);
+            if (!existed)
+            {
+                char *msg;
+                if (!NULL_OR_EMPTY(pkg_arch) && !NULL_OR_EMPTY(pkg_arch))
+                {
+                    xasprintf(&msg, "Package '%s-%s [%s]' would be installed\n", pkg_name, pkg_arch, pkg_ver);
```

Comment:
Memory for `msg` is allocated on the heap with `xasprintf` but I can't see `msg` freed any where.

Reply:
Yup. Another good catch!

Reply:
Actually, `msg` is inserted into a `StringMap` which is then destroyed (incl. these strings).

Reply:
Aha, `StringMap` takes ownership. Makes sense. :+1: